File: src/cmdbox/services/run_service.py
import json
from typing import Callable

from cmdbox.models import Command, Profile
from cmdbox.repositories.command_repository import CommandRepository
from cmdbox.repositories.history_repository import HistoryRepository
from cmdbox.repositories.profile_repository import ProfileRepository
from cmdbox.resolve.resolver import Resolver
from cmdbox.resolve.type_defs import ResolveResult, RefKind
from cmdbox.runtime.executor import Executor, RunContext
from cmdbox.runtime.results import ExecutionResult
from cmdbox.settings.models import Settings


class RunService:
    """
    Provides services for executing and previewing commands.

    This class manages the execution of commands and their previews. It interacts with
    a command repository, a resolver for processing commands, and an executor for
    executing commands. The `run` method is used to execute a command, while the `preview`
    method enables retrieving a resolved result of a command without executing it.

    Attributes:
        repo (CommandRepository): The repository that stores and provides access to
            command definitions based on their alias.
        resolver (Resolver): Encapsulates the logic required to resolve a command
            template into its executable form.
        executor (Executor): Responsible for executing resolved commands and
            returning the results.
    """

    def __init__(
        self,
        repo: CommandRepository,
        resolver: Resolver,
        executor: Executor,
        profile_repo: ProfileRepository,
        history_repo: HistoryRepository | None = None,
        get_settings: Callable[[], Settings] | None = None,
    ) -> None:
        self._repo = repo
        self._resolver = resolver
        self._executor = executor
        self._profile_repo = profile_repo
        self._history_repo = history_repo
        self._get_settings = get_settings

    def _resolve_profile(self, profile: str | None) -> Profile | None:
        return self._profile_repo.get_by_name(profile) if profile else None

    def run(
        self,
        command_alias: str,
        ctx: RunContext | None = None,
        runtime_vars: dict[str, str] | None = None,
        profile: str | None = None,
    ) -> ExecutionResult:
        """
        Executes a command based on the given alias.

        This method retrieves a command associated with the provided alias, resolves its
        template using a resolver, and then executes the resolved command text. Execution
        context stored on the command (cwd, shell, env, timeout) is merged with any
        runtime-supplied context, with runtime values taking precedence.

        Args:
            command_alias (str): The alias for the command to be executed.
            ctx (RunContext | None): The context for running the command.
            runtime_vars (dict[str, str] | None): Runtime variables to be used during command execution.
            profile (str | None): The name of the profile to use for command execution.

        Returns:
            ExecutionResult: The result of executing the command.
        """
        resolved_profile = self._resolve_profile(profile)
        cmd = self._repo.get_by_alias(command_alias, profile=resolved_profile)
        resolved_cmd = self._resolver.resolve(cmd.template, runtime_vars=runtime_vars)
        ctx = self.build_context(cmd, ctx)
        result = self._executor.run(resolved_cmd.text, ctx=ctx)
        self._repo.record_use(cmd.id)
        self.record_history(
            alias=command_alias,
            template=cmd.template,
            resolved=resolved_cmd.text,
            runtime_vars=runtime_vars,
            exit_code=result.exit_code,
            profile=resolved_profile,
        )
        self.record_variable_profile_use(resolved_cmd)
        return result

    def preview(
        self,
        command_alias: str,
        runtime_vars: dict[str, str] | None = None,
        ctx: RunContext | None = None,
        profile: str | None = None,
    ) -> tuple[ResolveResult, RunContext | None]:
        """
        Retrieves and resolves a command template based on its alias.

        This method takes a command alias, retrieves the associated command template,
        and resolves it to produce a complete command configuration or data structure.
        This resolved template is then returned as a `ResolveResult`.

        Args:
            command_alias (str): The alias of the command to be resolved.
            runtime_vars (dict[str, str] | None): Runtime variables to be used during command resolution.
            ctx (RunContext | None): The context in which the command is being executed.
            profile (str | None): The name of the profile to be used for command resolution. Defaults to None.

        Returns:
            tuple[ResolveResult, RunContext | None]: The resolved result of the command template and the
            effective context.
        """
        resolved_profile = self._resolve_profile(profile)
        cmd = self._repo.get_by_alias(command_alias, profile=resolved_profile)
        resolved = self._resolver.resolve(cmd.template, runtime_vars=runtime_vars)
        effective_ctx = self.build_context(cmd, ctx)
        return resolved, effective_ctx

    def record_history(
        self,
        *,
        alias: str,
        template: str,
        resolved: str,
        profile: Profile | None,
        runtime_vars: dict[str, str] | None = None,
        exit_code: int | None = None,
    ) -> None:
        """
        Records command execution history if the history feature is enabled and
        properly configured. This function interacts with an underlying
        repository to store details about the executed command including alias,
        template, resolved string, runtime variables, and exit code.

        Args:
            alias (str): The alias or the name of the command executed.
            template (str): The template string representing the command.
            resolved (str): The fully resolved string of the executed command.
            profile (Profile): The profile used for command execution.
            runtime_vars (dict[str, str] | None): Runtime variables used during
                the execution of the command. Defaults to None.
            exit_code (int | None): Exit code that indicates the result of the
                command execution. Defaults to None.
        """
        if not self._history_repo or not self._get_settings:
            return
        settings = self._get_settings()
        if not settings.history.enabled:
            return
        self._history_repo.record(
            alias=alias,
            template=template,
            resolved=resolved,
            variables_used=runtime_vars or None,
            exit_code=exit_code,
            limit=settings.history.limit_per_command,
            profile=profile,
        )

    def record_variable_profile_use(self, resolved_cmd: ResolveResult) -> None:
        """
        Records the usage of a variable profile if a stored variable is detected in
        the resolution trace of the provided command.

        This method inspects the trace of the provided `resolved_cmd` to determine
        if a variable with a reference kind of `VARIABLE` and a source of `stored`
        was used. If such a variable is found, the method retrieves the active
        variable profile ID from the profile repository and records its usage.

        Args:
            resolved_cmd: The resolution result containing the trace of steps
                executed and their metadata.

        """
        used_stored_variable = any(
            step.kind == RefKind.VARIABLE and step.source == "stored"
            for step in resolved_cmd.trace
        )
        if used_stored_variable:
            active_variable_profile_id = self._profile_repo.get_state().active_variable_profile_id
            self._profile_repo.record_use(active_variable_profile_id)

    def collect_missing_vars(
        self,
        command_alias: str,
        runtime_vars: dict[str, str] | None = None,
        profile: str | None = None,
    ) -> list[str]:
        resolved_profile = self._resolve_profile(profile)
        cmd = self._repo.get_by_alias(command_alias, profile=resolved_profile)
        missing = self._resolver.collect_missing_vars(
            cmd.template, runtime_vars=runtime_vars
        )
        return missing

    def build_context(self, cmd: Command, runtime_ctx: RunContext | None = None) -> RunContext | None:
        """
        Builds and returns a runtime execution context by consolidating information from the command and
        the current runtime context.

        The method integrates the environment variables, working directory, shell, timeout, output
        capture settings, verbosity, and other runtime parameters to form a comprehensive context. If
        certain properties are not provided in the runtime context, it falls back to values specified
        in the command or default application settings.

        Args:
            cmd (Command): The command object containing configuration and parameters for execution.
            runtime_ctx (RunContext | None): The optional runtime context that can override specific
                properties of the command settings.

        Returns:
            RunContext | None: A consolidated execution context containing the merged settings to be
                used during command execution.
        """
        stored_env = json.loads(cmd.env) if cmd.env else {}
        runtime_env = getattr(runtime_ctx, "env", None) or {}
        merged_env = {**stored_env, **runtime_env} or None

        cwd = (
          runtime_ctx.cwd if runtime_ctx and runtime_ctx.cwd is not None else None
        ) or cmd.cwd

        settings = self._get_settings()

        if runtime_ctx and runtime_ctx.shell is not None:
            shell = runtime_ctx.shell
        elif cmd.shell is not None:
            shell = cmd.shell
        else:
            shell = settings.execution_settings.default_shell

        if runtime_ctx and runtime_ctx.timeout is not None:
            timeout = runtime_ctx.timeout
        else:
            timeout = cmd.timeout

        if runtime_ctx and runtime_ctx.capture is not None:
            capture = runtime_ctx.capture
        else:
            capture = settings.execution_settings.capture_output

        if runtime_ctx and runtime_ctx.verbose is not None:
            verbose = runtime_ctx.verbose
        else:
            verbose = settings.execution_settings.default_verbose

        emit = runtime_ctx.emit if runtime_ctx else False

        return RunContext(
            cwd=cwd,
            shell=shell,
            env=merged_env,
            timeout=timeout,
            capture=capture,
            emit=emit,
            verbose=verbose,
        )

File: internal/services/run_service.go
package services

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"maps"

	"github.com/PhantomLambSoft/CmdBox/internal/models"
	"github.com/PhantomLambSoft/CmdBox/internal/repository"
	"github.com/PhantomLambSoft/CmdBox/internal/resolve"
	"github.com/PhantomLambSoft/CmdBox/internal/runtime"
	"github.com/PhantomLambSoft/CmdBox/internal/settings"
)

// SettingsProvider defines an interface for allowing SettingsService to be mocked for testing without dealing with
// the I/O functionality that comes with SettingsService.
type SettingsProvider interface {
	Get() settings.Settings
}

type RuntimeOverrides struct {
	Cwd     string
	Env     map[string]string
	Capture *bool
	Shell   *string
	Timeout *int
	Emit    bool
	Verbose *bool
}

type RunService struct {
	cmdRepo          repository.CommandRepository
	resolver         resolve.Resolver
	executor         runtime.Executor
	profileRepo      repository.ProfileRepository
	historyRepo      repository.HistoryRepository
	settingsProvider SettingsProvider
}

type RunServiceConfig struct {
	CmdRepo          repository.CommandRepository
	Resolver         resolve.Resolver
	Executor         runtime.Executor
	ProfileRepo      repository.ProfileRepository
	HistoryRepo      repository.HistoryRepository
	SettingsProvider SettingsProvider
}

func NewRunService(config RunServiceConfig) *RunService {
	return &RunService{
		cmdRepo:          config.CmdRepo,
		resolver:         config.Resolver,
		executor:         config.Executor,
		profileRepo:      config.ProfileRepo,
		historyRepo:      config.HistoryRepo,
		settingsProvider: config.SettingsProvider,
	}
}

func (s *RunService) resolveProfile(profileName *string) (*models.Profile, error) {
	if profileName != nil {
		if *profileName != "" {
			profile, err := s.profileRepo.GetByName(*profileName)
			if err != nil {
				return nil, fmt.Errorf("getting profile: %w", err)
			}
			return profile, nil
		}
	}
	return nil, nil
}

func (s *RunService) Run(
	cmdAlias string, runtimeOverrides RuntimeOverrides, runtimeVars map[string]string, profileName *string,
) (runtime.ExecutionResult, error) {
	profile, err := s.resolveProfile(profileName)
	if err != nil {
		return runtime.ExecutionResult{}, err
	}
	var profileID *uint
	if profile != nil {
		profileID = &profile.ID
	}

	command, err := s.cmdRepo.GetByAlias(cmdAlias, profileID)
	if err != nil {
		return runtime.ExecutionResult{}, fmt.Errorf("getting command: %w", err)
	}

	resolveResult, err := s.resolver.Resolve(command.Template, nil, runtimeVars)
	if err != nil {
		return runtime.ExecutionResult{}, fmt.Errorf("resolving command: %w", err)
	}
	resolvedCommand := resolveResult.Text

	effectiveCtx, err := s.buildContext(*command, runtimeOverrides)
	if err != nil {
		return runtime.ExecutionResult{}, fmt.Errorf("building context: %w", err)
	}

	// TODO: when effectiveCtx.Emit is true, executor.Run calls os.Exit(0) directly instead of returning,
	// which kills the calling process (untestable in-process) and would leave executionResult nil -
	// panicking below on `*executionResult` - if that exit path were ever removed. Consider having the
	// executor return a normal result/sentinel for emit instead of exiting itself.
	executionResult, err := s.executor.Run(resolvedCommand, effectiveCtx)
	if err != nil {
		return runtime.ExecutionResult{}, fmt.Errorf("running command: %w", err)
	}

	if err = s.cmdRepo.RecordUse(command.ID); err != nil {
		slog.Warn("failed to record command use", "alias", cmdAlias, "error", err)
	}

	historyInput := RecordHistoryInput{
		Alias:       cmdAlias,
		Template:    command.Template,
		Resolved:    resolvedCommand,
		ProfileID:   profileID,
		RuntimeVars: runtimeVars,
		ExitCode:    executionResult.ExitCode,
	}

	if _, err = s.recordHistory(historyInput); err != nil {
		slog.Warn("failed to record history", "alias", cmdAlias, "error", err)
	}

	if err = s.recordVariableProfileUse(resolveResult); err != nil {
		slog.Warn("failed to record variable profile use", "alias", cmdAlias, "error", err)
	}

	return *executionResult, nil
}

func (s *RunService) Preview(
	cmdAlias string, runtimeOverrides RuntimeOverrides, runtimeVars map[string]string, profileName *string,
) (resolve.Result, runtime.RunContext, error) {
	profile, err := s.resolveProfile(profileName)
	if err != nil {
		return resolve.Result{}, runtime.RunContext{}, err
	}
	var profileID *uint
	if profile != nil {
		profileID = &profile.ID
	}

	command, err := s.cmdRepo.GetByAlias(cmdAlias, profileID)
	if err != nil {
		return resolve.Result{}, runtime.RunContext{}, fmt.Errorf("getting command: %w", err)
	}

	resolveResult, err := s.resolver.Resolve(command.Template, nil, runtimeVars)
	if err != nil {
		return resolve.Result{}, runtime.RunContext{}, fmt.Errorf("resolving command: %w", err)
	}

	effectiveCtx, err := s.buildContext(*command, runtimeOverrides)
	if err != nil {
		return resolve.Result{}, runtime.RunContext{}, fmt.Errorf("building context: %w", err)
	}

	return resolveResult, effectiveCtx, nil
}

type RecordHistoryInput struct {
	Alias       string
	Template    string
	Resolved    string
	ProfileID   *uint
	RuntimeVars map[string]string
	ExitCode    int
}

func (s *RunService) recordHistory(input RecordHistoryInput) (*models.CommandHistory, error) {
	currentSettings := s.settingsProvider.Get()
	if !currentSettings.History.Enabled {
		return nil, nil
	}
	if s.historyRepo == nil {
		return nil, nil
	}

	var variablesUsed *string
	if len(input.RuntimeVars) > 0 {
		marshalled, err := json.Marshal(input.RuntimeVars)
		if err != nil {
			return nil, fmt.Errorf("marshaling runtime vars: %w", err)
		}
		s := string(marshalled)
		variablesUsed = &s
	}

	historyEntry := repository.HistoryEntry{
		Alias:          input.Alias,
		Template:       input.Template,
		Resolved:       input.Resolved,
		VariablesUsed:  variablesUsed,
		ExitCode:       &input.ExitCode,
		RetentionLimit: &currentSettings.History.LimitPerCommand,
		ProfileID:      input.ProfileID,
	}

	history, err := s.historyRepo.Record(historyEntry)
	if err != nil {
		return nil, fmt.Errorf("recording history: %w", err)
	}

	return history, nil
}

func (s *RunService) recordVariableProfileUse(resolvedCommand resolve.Result) error {
	usedStoredVar := false
	for _, step := range resolvedCommand.Trace {
		if step.Kind == resolve.RefKindVariable && step.Source == "stored" {
			usedStoredVar = true
			break
		}
	}
	if !usedStoredVar {
		return nil
	}

	profile, err := s.profileRepo.GetActiveVariableProfile()
	if err != nil {
		return fmt.Errorf("getting active variable profile: %w", err)
	}
	return s.profileRepo.RecordUse(*profile)
}

// CollectMissingVars identifies variables inside a command template that are not provided in the supplied runtimeVars map.
func (s *RunService) CollectMissingVars(cmdAlias string, runtimeVars map[string]string, profileName *string) ([]string, error) {
	profile, err := s.resolveProfile(profileName)
	if err != nil {
		return []string{}, fmt.Errorf("resolving profile: %w", err)
	}
	var profileID *uint
	if profile != nil {
		profileID = &profile.ID
	}
	cmd, err := s.cmdRepo.GetByAlias(cmdAlias, profileID)
	if err != nil {
		return []string{}, fmt.Errorf("getting command: %w", err)
	}
	missing, err := s.resolver.CollectMissingVars(cmd.Template, runtimeVars)
	if err != nil {
		return nil, fmt.Errorf("collecting missing vars: %w", err)
	}
	return missing, nil
}

// BuildContext constructs a runtime.RunContext by combining runtime override configurations with stored command
// configurations and default values taken from settings.
func (s *RunService) buildContext(cmd models.Command, runtimeOverrides RuntimeOverrides) (runtime.RunContext, error) {
	var err error
	envs := make(map[string]string)
	if cmd.Env != nil {
		envs, err = parseEnv(*cmd.Env)
		if err != nil {
			return runtime.RunContext{}, fmt.Errorf("parsing stored env: %w", err)
		}
	}
	if runtimeOverrides.Env != nil {
		maps.Copy(envs, runtimeOverrides.Env) // merge runtime envs into stored envs under envs variable
	}

	currentSettings := s.settingsProvider.Get()

	var cwd = ""
	if runtimeOverrides.Cwd != "" {
		cwd = runtimeOverrides.Cwd
	} else if cmd.Cwd != nil {
		cwd = *cmd.Cwd
	}

	var shell *string
	if runtimeOverrides.Shell != nil {
		shell = runtimeOverrides.Shell
	} else if cmd.Shell != nil {
		shell = cmd.Shell
	} else {
		shell = &currentSettings.ExecutionSettings.DefaultShell
	}

	var timeout = 0
	if runtimeOverrides.Timeout != nil {
		timeout = *runtimeOverrides.Timeout
	} else if cmd.Timeout != nil {
		timeout = *cmd.Timeout
	}

	var capture *bool
	if runtimeOverrides.Capture != nil {
		capture = runtimeOverrides.Capture
	} else {
		capture = &currentSettings.ExecutionSettings.CaptureOutput
	}

	var verbose *bool
	if runtimeOverrides.Verbose != nil {
		verbose = runtimeOverrides.Verbose
	} else {
		verbose = &currentSettings.ExecutionSettings.DefaultVerbose
	}

	emit := runtimeOverrides.Emit

	return runtime.RunContext{
		Cwd:     cwd,
		Env:     envs,
		Shell:   *shell,
		Timeout: timeout,
		Capture: *capture,
		Verbose: *verbose,
		Emit:    emit,
	}, nil
}
